### PR TITLE
Closes #206 | Add Dataloader SleukRith Set

### LINKED ISSUE #206: Create dataset loader for SleukRith Set
Dataloader name: `sleukrith_ocr/sleukrith_ocr.py`
DataCatalogue: http://seacrowd.github.io/seacrowd-catalogue/card.html?sleukrith_ocr

|     Dataset| sleukrith_ocr |
|-------------|---|
| Description | SleukRith Set is the first dataset specifically created for Khmer palm leaf manuscripts. The dataset consists of annotated data from 657 pages of digitized palm leaf manuscripts which are selected arbitrarily from a large collection of existing and also recently digitized images. The dataset contains three types of data: isolated characters, words, and lines. Each type of data is annotated with the ground truth information which is very useful for evaluating and serving as a training set for common document analysis tasks such as character/text recognition, word/line segmentation, and word spotting. |
| Subsets | - |
| Languages | khm |
| Tasks | Optical Character Recognition |
| License | Unknown (unknown) |
| Homepage | https://github.com/donavaly/SleukRith-Set |
| HF URL | - |
| Paper URL | https://dl.acm.org/doi/10.1145/3151509.3151510 |

### PR DESCRIPTION
**Pull request by akhdanfadh**
Closes #206

Sorry if this PR is made before any final decision in #206, I just want to rush things.

Some possible discussions:

1. Add another task for OCR (see #555). After that one is merged, I will remove the corresponding file here.
2. Used third-party libraries to download the data, i.e., `pip install gdown`. I am aware that I should made a separate PR for this, just want to put this here first for reviewers to discuss the implementation. Related discussion is on #206.
3. I store the downloaded data in `data/sleukrith_ocr/`, thus the updated `.gitignore`. Again, should be on a separate PR, right?
4. The raw image dataset is in binary. I convert the image data as numpy arrays, and then save the image as a file so I can store the image path for the schema (see `_generate_examples()`). What do you think about that? Should I just store the image as numpy arrays instead?

### Checkbox
- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `seacrowd/sea_datasets/{my_dataset}/{my_dataset}.py` (please use only lowercase and underscore for dataset folder naming, as mentioned in dataset issue) and its `__init__.py` within `{my_dataset}` folder.
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_LOCAL`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py` or `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py --subset_id {subset_name_without_source_or_seacrowd_suffix}`.
- [ ] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

### CODE REVIEW

File: seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+                    "image_path": example[0],
+                    "label": example[1],
+                }
+
+        elif self.config.schema == _SEACROWD_SCHEMA:
+            for idx, example in enumerate(zip(image_paths, labels)):
+                yield idx, {
+                    "id": str(idx),
+                    "image_paths": [example[0]],
+                    # "texts": None,
```
Comment (holylovenia) [inline seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py:240]: Uncomment this.

File: seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+                    "metadata": {
+                        # "context": None,
```
Comment (holylovenia) [inline seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py:242]: Uncomment this line too.

File: seacrowd/utils/constants.py [hunk 3]
```diff
@@ -125,6 +125,7 @@ class Tasks(Enum):
     IMAGE_CAPTIONING = "IC"
     VISUAL_QUESTION_ANSWERING = "VQA"
     SIGN_LANGUAGE_RECOGNITION = "SLR"
+    OPTICAL_CHARACTER_RECOGNITION = "OCR"
```
Comment (holylovenia) [inline seacrowd/utils/constants.py:128]: This task has to be added in a separate PR.

File: seacrowd/utils/constants.py [hunk 4]
```diff
@@ -282,6 +283,7 @@ class Licenses(Enum):
     Tasks.VISUAL_QUESTION_ANSWERING: "IMQA",
     Tasks.IMAGE_CAPTIONING: "IMTEXT",
     Tasks.SIGN_LANGUAGE_RECOGNITION: "IMTEXT",
+    Tasks.OPTICAL_CHARACTER_RECOGNITION: "IMTEXT",
```
Comment (holylovenia) [inline seacrowd/utils/constants.py:286]: This task has to be added in a separate PR.

**Review by holylovenia — CHANGES_REQUESTED** [review]
Hi @akhdanfadh, it works well on my side. I have some suggestions:

1. For the additional packages (i.e., `PIL` and `gdown`), we could use `try-except` to ask the users to install the packages without needing to add them to the `requirements.txt`. Example:
```
try:
    import yourModule
except ImportError:
    raise Exception(“…”)
```

**Comment by akhdanfadh** [comment]
@holylovenia Resolve some changes and added additional package handling.

> This task has to be added in a separate PR.

I'll make the new PR once everything on the dataloader part is done and reviewed.

**Review by holylovenia — APPROVED** [review]
LGTM! Aside from needing to remove the changes in `requirements.txt` and `constants.py`, everything else works well on my side. Thanks @akhdanfadh! Let's wait for @sabilmakbar's review.

**Comment by akhdanfadh** [comment]
IDK why `constants.py` is still listed as a changed file here even tho I already changed it by running `git checkout upstream/master -- seacrowd/utils/constants.py`. I think I will delete the file manually then.

I think I also need to make a new PR for adding `data/` in `.gitignore`, right? @holylovenia

**Comment by holylovenia** [comment]
> I think I also need to make a new PR for adding `data/` in `.gitignore`, right? @holylovenia

Yes yes, I'll approve that PR once you create it.

**Comment by akhdanfadh** [comment]
> Yes yes, I'll approve that PR once you create it.

#658 and files removed. This PR should be done on my part.

**Review by holylovenia — APPROVED** [review]
Thanks @akhdanfadh! Let's wait for @sabilmakbar's review.

File: seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+
+_LICENSE = Licenses.UNKNOWN.value
+
+_LOCAL = False
+
+_URLS = {
+    "sleukrith-set": {
+        "images": "https://drive.google.com/uc?export=download&id=19JIxAjjXWuJ7mEyUl5-xRr2B8uOb-GKk",  # 1GB
+        "annotated-data": "https://drive.google.com/uc?export=download&id=1Xi5ucRUb1e9TUU-nv2rCUYv2ANVsXYDk",  # 11.7MB
+    },
```
Comment (sabilmakbar) [inline seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py:89]: looks like this is unused anywhere on the dataloader, is it necessary>

File: seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+    comprising of 657 pages of Khmer palm leaf manuscripts randomly selected
+    from various collections whose quality and digitization method are variable.
+    The dataset contains three types of data: isolated characters, words, and
+    lines. Each type of data is annotated with the ground truth information
+    which is very useful for evaluating and serving as a training set for common
+    document analysis tasks such as character/text recognition, word/line
+    segmentation, and word spotting. In order to serve as a base line, the
+    result of an evaluation study of Khmer isolated character recognition that
+    we have conducted on SleukRith Set using Convolutional Neural Network is
+    also presented.},
```
Comment (sabilmakbar) [inline seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py:54]: the abstract is quite long and unnecessary in citation. Should we remove it?

File: seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+    booktitle = {Proceedings of the 4th International Workshop on Historical Document Imaging and Processing},
+    pages = {1-6},
+    numpages = {6},
+    keywords = {palm leaf manuscript, ground truth, annotated dataset, Handwritten document analysis},
```
Comment (sabilmakbar) [inline seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py:58]: same goes to this `keywords` field

File: seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+    ]
+
+    DEFAULT_CONFIG_NAME = f"{_DATASETNAME}_source"
+
+    def _info(self) -> datasets.DatasetInfo:
+        if self.config.schema == "source":
+            features = datasets.Features(
+                {
+                    "image_path": datasets.Value("string"),
+                    "label": datasets.ClassLabel(names=[i for i in range(111)]),
```
Comment (sabilmakbar) [inline seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py:136]: Does this dataset have label metadata for each integer value from 0-110?

File: seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+            citation=_CITATION,
+        )
+
+    def module_exists(self, module_name):
+        try:
+            mod = __import__(module_name)
+        except ImportError:
+            return False
+        else:
+            return True
```
Comment (sabilmakbar) [inline seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py:156]: prob we need to add this into SEACrowd helpers (or add importing module that has global scope var), but we can leave this for now.

**Review by sabilmakbar — COMMENTED** [review]
few comments, the dataloader works fine

**Comment by holylovenia** [comment]
A friendly reminder for @akhdanfadh to address @sabilmakbar's suggestions. 🙏

File: seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py (hunk 5 above)
Comment (akhdanfadh) [reply seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py:89]: You are correct that it is unused here. My way of thinking is that if we consider the global variable to be the dataset metadata, it is okay to put it there. Inline comment also works in that case.

This `sleukrith-set` key is the raw data (the whole image), and `isolated-characters` is the processed data (cropped image per char).

File: seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py (hunk 8 above)
Comment (akhdanfadh) [reply seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py:136]: No. Please refer to the discussion in the datasheet, especially Q1.

File: seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py (hunk 9 above)
Comment (akhdanfadh) [reply seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py:156]: Done removing `mod` variable name.

Adding to SEACrowd helpers is optional, just let me know if you want this and I'll make a PR for that.

**Comment by akhdanfadh** [comment]
@sabilmakbar Done, please check my replies.

File: seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py (hunk 5 above)
Comment (sabilmakbar) [reply seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py:70]: hmm prob an inline comment for clarifying this would be enough, like this one:

> This sleukrith-set key is the raw data (the whole image), and isolated-characters is the processed data (cropped image per char).

And it's okay if we put it in the `_URL` and note down the ones we're using is `isolated-characters` key instead

File: seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py (hunk 8 above)
Comment (sabilmakbar) [reply seacrowd/sea_datasets/sleukrith_ocr/sleukrith_ocr.py:117]: okay, thanks for pointing it out. prob we can note that the label meta wasn't given as inline comment before L136

**Comment by akhdanfadh** [comment]
@sabilmakbar Done with the inline comments.

**Comment by sabilmakbar** [comment]
lgtm, thanks @akhdanfadh!
